url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/resid/SID.java

```diff
@@ -0,0 +1,1321 @@
+			state.envelope_state[i] = voice[i].envelope.state;
+			state.hold_zero[i] = voice[i].envelope.hold_zero;
+		}
+
+		return state;
+	}
+
+	/**
+	 * Write state.
+	 * 
+	 * @param state
+	 */
+	public void write_state(final State state) {
+		int i;
+
+		for (i = 0; i <= 0x18; i++) {
+			write(i, state.sid_register[i]);
+		}
+
+		bus_value = state.bus_value;
+		bus_value_ttl = state.bus_value_ttl;
+
+		for (i = 0; i < 3; i++) {
+			voice[i].wave.accumulator = state.accumulator[i];
+			voice[i].wave.shift_register = state.shift_register[i];
+			voice[i].envelope.rate_counter = state.rate_counter[i];
+			voice[i].envelope.rate_period = state.rate_counter_period[i];
+			voice[i].envelope.exponential_counter = state.exponential_counter[i];
+			voice[i].envelope.exponential_counter_period = state.exponential_counter_period[i];
+			voice[i].envelope.envelope_counter = state.envelope_counter[i];
+			voice[i].envelope.state = state.envelope_state[i];
+			voice[i].envelope.hold_zero = state.hold_zero[i];
+		}
+	}
+
+	/**
+	 * Enable filter.
+	 * 
+	 * @param enable
+	 */
+	public void enable_filter(boolean enable) {
+		filter.enable_filter(enable);
+	}
+
+	/**
+	 * Enable external filter.
+	 * 
+	 * @param enable
+	 */
+	public void enable_external_filter(boolean enable) {
+		extfilt.enable_filter(enable);
+	}
+
+	/**
+	 * I0() computes the 0th order modified Bessel function of the first kind.
+	 * This function is originally from resample-1.5/filterkit.c by J. O. Smith.
+	 * 
+	 * @param x
+	 * @return
+	 */
+	protected double I0(double x) {
+		// Max error acceptable in I0.
+		final double I0e = 1e-6;
+
+		double sum, u, halfx, temp;
+		int n;
+
+		sum = u = n = 1;
+		halfx = x / 2.0;
+
+		do {
+			temp = halfx / n++;
+			u *= temp * temp;
+			sum += u;
+		} while (u >= I0e * sum);
+
+		return sum;
+	}
+
+	/**
+	 * Setting of SID sampling parameters.
+	 * <P>
+	 * Use a clock freqency of 985248Hz for PAL C64, 1022730Hz for NTSC C64. The
+	 * default end of passband frequency is pass_freq = 0.9*sample_freq/2 for
+	 * sample frequencies up to ~ 44.1kHz, and 20kHz for higher sample
+	 * frequencies.
+	 * <P>
+	 * For resampling, the ratio between the clock frequency and the sample
+	 * frequency is limited as follows: 125*clock_freq/sample_freq < 16384 E.g.
+	 * provided a clock frequency of ~ 1MHz, the sample frequency can not be set
+	 * lower than ~ 8kHz. A lower sample frequency would make the resampling
+	 * code overfill its 16k sample ring buffer.
+	 * <P>
+	 * The end of passband frequency is also limited: pass_freq <=
+	 * 0.9*sample_freq/2
+	 * <P>
+	 * E.g. for a 44.1kHz sampling rate the end of passband frequency is limited
+	 * to slightly below 20kHz. This constraint ensures that the FIR table is
+	 * not overfilled.
+	 * 
+	 * @param clock_freq
+	 * @param method
+	 * @param sample_freq
+	 * @param pass_freq
+	 * @param filter_scale
+	 * @return
+	 */
+	public boolean set_sampling_parameters(double clock_freq,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 740-853 into a separate method.